man/mcp-sample.1
.\" Copyright (c) 2019-2024 Miki Hermann & Gernot Salzer
.TH mcp-sample 1 "2024-04-13" "1.04" "MCP System"
.
.SH NAME
mcp-sample -- sampling uniformly at random from a large data file
.
.SH SYNOPSIS
.B mcp-sample
.RI [\| "OPTION" "\|]\|.\|.\|."
.
.SH DESCRIPTION
.PP
Data is sampled from input file uniformly at random according to a
confidence interval or an error bound, similarly to opinion polls,
with maintained ratio of concept values.
.
.SH OPTIONS
.TP
\fB\-i\fR, \fB\-\-input\fI input-file
Input file from which a sample will be extracted.
.IP
Default: STDIN.
.
.TP
\fB\-o\fR, \fB\-\-output\fI output-file
Output file of sampled data.
.IP
Default: STDOUT or \fIbasename\fR of \fIinput-file\fR with
\_\fIsample\fR infix and \fIinput-file\fR suffix.
.
.TP
\fB\-\-pp\fR, \fB\-\-population\fR proportional | absolute
.IP
proportional | prop:
.br
The proportion in the sample of items with different concept values reflects
their proportion in the input data file. It makes calls to \fImcp-clean\fR
to select items for each concept value from the input file and
recursive calls to \fImcp-sample\fR with \fIabsolute\fR population parameter
to compute the sample from them.
.IP
absolute | abs:
.br
The sample is taken from the input data file as a whole without respecting
the proportions of concept values.
.IP
Default: proportional
.
.TP
\fB\-W\fR, \fB\-\-confidence\fR \fIFLOAT \fR or \fR \fIPERCENTAGE \fR
.IP
Confidence interval written either as a float between 0.001 and 0.2 or a
percentage with \fI%\fR sign between 0.1% and 20%. Equivalent to
double of \fIerror bound\fR. Cannot be specified together with
\fIerror bound\fR.
.IP
Default: 2.5%.
.
.TP
\fB\-B\fR, \fB\-\-bound\fR \fIFLOAT \fR or \fR \fIPERCENTAGE \fR
.IP
Error bound written either as a float between 0.0005 and 0.1 or a
percentage with \fI%\fR sign between 0.05% and 10%. Equivalent to half
of \fIconfidence interval\fR. Cannot be specified together with
\fIconfidence interval\fR.
.
.TP
\fB\-#\fR, \fB\-\-card\fR \fINUMBER \fR
.IP
Cardinality of the sample. Overrides confidence interval and error
bound.
.
.TP
\fB\-\-big\fR
.IP
Indication that the data file is too big (e.g., 100GB) to be held
entirely in the memory.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
.IP
Quiet mode.
.
.SH SEE ALSO
mcp-clean(1)
.
.SH BUGS
There are certanly some.
.
.SH AUTHORS
Miki Hermann <hermann@lix.polytechnique.fr>
.br
Gernot Salzer <gernot.salzer@tuwien.ac.at>


changelog.txt
1.04
- fixed a bug in mcp-check.cpp
- added LaTeX output
- mcp-split introduced
- post_prod introduced
- write formula to a file (DIMACS form)

1.03c
 - added print option DIMACS
 - corrected sorting bug when '--cooking raw'
 - corrected bug when '--strategy exact' and '--closure Horn'
 - index of group_of_matrix: int -> string
 - variables can be named in the matrix
 - transformation in C++
 - failure to open input or output file results in exit and not in exception
 - added options 'peek' and 'section' for printing matrices
 - when variables have own names, literals made readable
 - bug correction: 'continue' instead of 'return' when empty formula
   in sequential for bijunctive 
 - added mixed printing of formulas
 - added resolution with binary clause which contracts the target clause
   (l0 + l1) * (l0 + -l1 + c) => (l0 + l1) * (l0 + c)
 - added mcp-cnf producting exact CNF formula


src-noarch/mcp-sample.cpp
/**************************************************************************
 *                                                                        *
 *                                                                        *
 *	         Multiple Classification Project (MCP)                    *
 *                                                                        *
 *	Author:   Miki Hermann                                            *
 *	e-mail:   hermann@lix.polytechnique.fr                            *
 *	Address:  LIX (CNRS UMR 7161), Ecole Polytechnique, France        *
 *                                                                        *
 *	Author: Gernot Salzer                                             *
 *	e-mail: gernot.salzer@tuwien.ac.at                                *
 *	Address: Technische Universitaet Wien, Vienna, Austria            *
 *                                                                        *
 *	Version: all                                                      *
 *      File:    mcp-sample.cpp                                           *
 *                                                                        *
 *      Copyright (c) 2019 - 2025                                         *
 *                                                                        *
 * Output a sample from an input file.                                    *
 *                                                                        *
 *                                                                        *
 **************************************************************************/

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <random>
#include <csignal>
#include "mcp-defs.hpp"
#include "mcp-basics.hpp"	// for split
#include "mcp-tally.hpp"

using namespace std;

string version = NOARCH_VERSION;

string input  = STDIN;
string output = STDOUT;
ifstream infile;
ofstream outfile;
ofstream routfile;
string tpath = "/tmp/mcp-tmp-";

string confidence = "2.5%";
double conf;
string error_bound;
string sample_card;
string proportion;
double prop;
string concept_column;
size_t ccol;		// concept column value
bool quiet = false;
bool big = false;

size_t number_of_lines = 0;
size_t sample_size = 0;

enum Population : char {proportional = 0, absolute = 1};
Population population = proportional;
enum Style : char {NOTHING = 0,
		   CONFIDENCE = 1,
		   BOUND = 2,
		   CARDINALITY = 3,
		   PROPORTION = 4};
Style style = NOTHING;

//+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

// reads the input parameters
void read_arg (int argc, char *argv[]) {
  int argument = 1;
  while (argument < argc) {
    string arg = argv[argument];
    if (arg == "--input"
	|| arg == "-i") {
      if (argument < argc-1) {
	input = argv[++argument];
      } else
	cerr << "+++ no input file selected" << endl;
    } else if (arg == "--output"
	       || arg == "-o") {
      if (argument < argc-1) {
	output = argv[++argument];
      } else
	cerr << "+++ no output file selected, revert to default" << endl;
    } else if (arg == "--population"
	       || arg == "--pop"
	       || arg == "--pp") {
      if (argument < argc-1) {
	arg = argv[++argument];
	if (arg == "proportional"
	    || arg == "prop")
	  population = proportional;
	else if (arg == "absolute"
		 || arg == "abs"
		 || arg == "a"
		 || arg == "A")
	  population = absolute;
	else
	  cerr << "+++ unknown population option " << arg << endl;
      } else
	cerr << "+++ no population option selected, revert to default" << arg << endl;
    } else if (arg == "--quiet"
	       || arg == "-q") {
      quiet = true;
    } else if (arg == "--confidence"
	       || arg == "--interval"
	       || arg == "-W"
	       || arg == "-w") {
      if (argument < argc-1) {
	confidence = argv[++argument];
	style = CONFIDENCE;
      } else
	cerr << "+++ no confidence interval selected, revert to default" << arg << endl;
    } else if (arg == "--bound"
	       || arg == "--error"
	       || arg == "-B"
	       || arg == "-b") {
      if (argument < argc-1) {
	error_bound = argv[++argument];
	style = BOUND;
      } else
	cerr << "+++ no error bound selected, revert to default" << arg << endl;
    } else if (arg == "--cardinality"
	       || arg == "--card"
	       || arg == "-#") {
      if (argument < argc-1) {
	sample_card = argv[++argument];
	style = CARDINALITY;
      } else
	cerr << "+++ no cardinality selected, revert to default" << arg << endl;
    } else if (arg == "--proportion"
	       || arg == "--prop") {
      if (argument < argc-1) {
	proportion = argv[++argument];
	style = PROPORTION;
      } else
	cerr << "+++ no proportion selected, revert to default" << arg << endl;
    } else if (arg == "--concept"
	       || arg == "-c") {
      if (argument < argc-1) {
	concept_column = argv[++argument];
      } else
	cerr << "+++ no concept column selected" << arg << endl;
    } else if (arg == "--big"
	       || arg == "--BIG") {
      big = true;
    } else
      cerr << "+++ unknown option " << arg << endl;
    ++argument;
  }
}

inline double string2double (const string &s) {
  return (s.back() == '%' ? stod(s.substr(0, s.length()-1)) * 0.01 : stod(s));
}

size_t sample_cardinality (const double &prop, const double &conf) {
  double val = (16.0 * prop * (1.0 - prop)) / (conf * conf);
  size_t sz = val;
  if (val > sz)
    sz++;
  return sz;
}

// test if sample size is correct and print out override information
inline void try_sample_size (const string &what) {
  try {
    sample_size = stoul(sample_card);
  } catch (invalid_argument err) {
    cerr << "+++ " << sample_card
	 << " is not a valid sample size"
	 << endl;
    exit(2);
  }
  if (!quiet)
    cerr << "+++ Sample cardinality " << sample_size
	 << " overrides " << what << endl;
  style = CARDINALITY;
}

// adjusts input parameters and open files
void adjust_and_open () {
  // input and output
  if (input != STDIN) {
    infile.open(input);
    if (infile.is_open())
      cin.rdbuf(infile.rdbuf());
    else {
      cerr << "+++ Cannot open input file " << input << endl;
      exit(2);
    }
  }

  if (output == STDOUT && input != STDIN) {
    auto pos = input.rfind('.');
    output = pos == string::npos
      ? input + "-sample.csv"
      : input.substr(0, pos) + "-sample" + input.substr(pos);
  }

  if (output != STDOUT) {
    if (quiet)
      outfile.open(output, ios::out | ios::app);
    else
      outfile.open(output, ios::out | ios::trunc);
    if (outfile.is_open())
      cout.rdbuf(outfile.rdbuf());
    else {
      cerr << "+++ Cannot open output file " << output << endl;
      exit(2);
    }
  }

  // coliding concept column
  if (!proportion.empty() && (!concept_column.empty())) {
    cerr << "+++ Both proportion and concept specified" << endl;
    exit(2);
  }

  if (concept_column.empty() && population == proportional) {
    cerr << "+++ Proportional population without concept column" << endl;
    exit(2);
  }

  if (! confidence.empty() && ! error_bound.empty()) {
    cerr << "+++ Both confidence interval and error bound specified" << endl;
    exit(2);
  } else if (! confidence.empty() && ! sample_card.empty()) {
    try_sample_size("confidence interval");
  } else if (! error_bound.empty() && ! sample_card.empty()) {
    try_sample_size("error bound");
  } else if (! error_bound.empty()) {
    try {
      conf = 2.0 * string2double(error_bound);
    } catch (invalid_argument err) {
      cerr << "+++ " << error_bound
	     << " is not a valid error bound"
	     << endl;
      exit(2);
    }
  } else if (! confidence.empty()) {
    try {
      conf = string2double(confidence);
    } catch (invalid_argument err) {
      cerr << "+++ " << confidence
	     << " is not a valid confidence interval"
	     << endl;
      exit(2);
    }
  } else if (! sample_card.empty()) {
    try {
      sample_size = stoul(sample_card);
    } catch (invalid_argument err) {
      cerr << "+++ " << sample_card
	     << " is not a valid sample size"
	     << endl;
      exit(2);
    }
  } else {
    cerr << "+++ Neither confidence interval, nor error bound, nore cardinality specified" << endl;
    exit(2);
  }

  if ((conf < 0.001 || conf > 0.2)
      && population == proportional
      && style != CARDINALITY) {
    cerr << "+++ Confidence interval reset to 2.5%" << endl;
    conf = 0.025;
  }

  if (concept_column.empty())
    try {
      if (proportion.empty())
	prop = 0.5;
      else if (proportion.back() == '%')
	prop = stod(proportion.substr(0, proportion.length()-1)) / 100.0;
      else
	prop = stod(proportion);
    } catch (invalid_argument err) {
      cerr << "+++ '" << proportion
	   << "' is not a valid proportion"
	   << endl;
      exit(2);
    }
  else {
    prop = 0.5;
    try {
      ccol = stoul(concept_column);
    } catch (invalid_argument err) {
      cerr << "+++ '" << concept_column
	     << "' after --concept is not a valid column number"
	     << endl;
	exit(2);
    }
  }
  if (sample_size == 0)
    sample_size = sample_cardinality(prop, conf);
}

void cleanup () {
  if (output != STDOUT)
    outfile.close();
  if (input != STDIN)
    infile.close();
}

// clear eof flags and seek to the beginning of input file
inline void reset_input () {
  cin.clear();
  infile.clear();
  cin.seekg(0);
  infile.seekg(0);
}

void first_pass () {
  string line;
  size_t lineno = 0;
  bool warning = false;
  
  while (getline(cin, line)) {
    lineno++;
    if (line.empty())
      continue;
    number_of_lines++;
    if (! concept_column.empty()) {
      vector<string> chunks = split(line, ", \t");
      if (ccol >= chunks.size()) {
	cerr << "++ concept column (" << ccol
	     << ") beyond item size (" << chunks.size()
	     << ") on input line "
	     << lineno
	     << endl;
	warning = true;
      }
    }
  }
  if (warning) {
    cerr << "+++ STOP because of input errors" << endl;
    exit(2);
  }
}

void second_pass () {
  string line;
  vector<string> data_lines, sample_lines;

  reset_input();
  while (getline(cin, line))
    data_lines.push_back(line);
  sample(data_lines.cbegin(), data_lines.cend(), back_inserter(sample_lines),
	 sample_size,
	 mt19937 {random_device{}()});

  for (const string &sline : sample_lines)
    cout << sline << endl;
}

void big_pass () {
  // install the random device
  random_device rd;
  static uniform_int_distribution<int> uni_dist(1, number_of_lines);
  static default_random_engine dre(rd());

  // generate the random numbers of chosen input lines and sort them
  vector<size_t> rand_nums;
  while (rand_nums.size() < sample_size) {
    size_t rnd = uni_dist(dre);
    auto it = find(rand_nums.cbegin(), rand_nums.cend(), rnd);
    if (it == rand_nums.cend())
      rand_nums.push_back(rnd);
  }
  sort(rand_nums.begin(), rand_nums.end());

  // search the sample lines in the big file
  string line;
  size_t linenum = 0;
  size_t pointer = 0;
  reset_input();
  while (getline(cin, line)) {
    if (line.empty())
      continue;
    linenum++;
    if (linenum == rand_nums[pointer]) {
      cout << line << endl;
      pointer++;
    }
  }
}

void erase_tmp () {
  const string temp_prefix = tpath + "mcp-tmp-";
  const string basename = "rm -f " + temp_prefix;
  const string erase_out  = basename + "*.out";
  const string erase_txt  = basename + "*.txt";
  system(erase_out.c_str());
  system(erase_txt.c_str());
}

// terminal handler: we erase the temporary files in case of a crash
void crash (int signal) {
  erase_tmp();
  cerr << endl << "\t*** Segmentation fault ***" << endl << endl;
  exit(signal);
}

// terminal handler: we erase the temporary files in case of an interrupt
void interrupt (int signal) {
  erase_tmp();
  cerr << endl << "\t*** Interrupt ***" << endl << endl;
  exit(signal);
}

//////////////////////////////////////////////////////////////////////////////

int main(int argc, char **argv) {
  version += "sample";
  cerr << "+++ version = " << version << endl;

  read_arg(argc, argv);
  adjust_and_open();
  // if (sample_size == 0)
  //   sample_size = sample_cardinality(prop, conf);
  signal(SIGSEGV, crash);
  signal(SIGINT, interrupt);

  if (population == absolute) {
    if (big || ! concept_column.empty())
      first_pass();
    if (big)
      big_pass();
    else
      second_pass();
  } else if (population == proportional) {
    tally(ccol);
    const time_t start_time = time(nullptr);
    const string basename = tpath + to_string(start_time);
    const string routname = basename + ".out";

    size_t real_size = 0;
    for (const auto &val : concept_items.percent) {
      // size_t section = (0.01 * val.second + 0.005 / percentage.size()) * sample_size;
      size_t section = (0.01 * val.second) * sample_size;
      cerr << "+++ section for " << val.first << " = " << section << endl;
      if (section == 0) {
	cerr << "+++ " + val.first + " section skipped" << endl;
	continue;
      }
      real_size += section;

      reset_input();
      routfile.open(routname);
      if (! routfile.is_open()) {
	cerr << "+++ Cannot open rout file " << routname << endl;
	exit(2);
      }

      string line;
      size_t lineno = 0;
      while (getline(cin, line)) {
	clear_line(++lineno, line);
	if (line.empty())
	  continue;
	uncomma_line(line);
	const vector<string> chunks = split(line, " \t");
	if (ccol >= chunks.size()) {
	  cerr << "+++ concept column out of range on line "
	       << lineno
	       << endl;
	  exit(2);
	}
	if (chunks[ccol] == val.first)
	  routfile << line << endl;
      }
      routfile.close();

      // cerr << "+++ recursive call to mcp-sample for " << val.first << endl;
      string rec_sample
	= "mcp-sample -i " + routname
	+ (big ? " --big" : "")
	// + " --big"
	+ " --population absolute -q -# " + to_string(section)
	+ " -o " + output;
      system(rec_sample.c_str());
      remove(routname.c_str());
    }
    sample_size = real_size;
  }
  cleanup();
  if (!quiet) {
    cerr << "+++ sample size = " << sample_size << endl
	 << "+++ sample written on " << output << endl;
  }
  return 0;
}

//////////////////////////////////////////////////////////////////////////////
